### DISCUSSION TITLE
[Idea] USB Serial protocol

### DISCUSSION THREAD

**Discussion by PaulFreund** [SynthstromAudible/DelugeFirmware#95] [Ideas]
After talking with @m-m-adams I understood that we should have enough USB descriptors to have (MPE) MIDI, Stereo Audio, USB File transfer and an additional Serial port all at the same time (or maybe even more). Since I know there is a lot of interrest to work on PC applications interfacing with the Deluge and there is also a lot of potential for interfacing in development I'd like to share an idea. 

* USB Serial connection with [JSON Lines](https://jsonlines.org/) protocol
* [cJSON](https://github.com/DaveGamble/cJSON) as parser and generator library
* Web applications can make use of the [Web Serial API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Serial_API)
* JavaScript applications will be able to directly use the JSON objects, for almost all other languages there are current libraries
* The protocol would be a simple { key: ..., value: ...} pair that can be easily routed to the relevant functions

Some functionality to implement with this that comes to mind:

* Serial logging output (mainly for debugging purposes)
* Integration testing automation (send key presses to the Deluge and check state variables) 
* Reading memory (maybe even export of dump files for offline debugging is possible?)
* Exchanging runtime values for various purposes (for example Synthesizer tuning etc.)
* Many more

Of course this can also be achieved in other ways like MIDI SysEx so I'd like to hear feedback which way is preferred and for what reasons :)

**Comment by litui** [comment]
Whatever we do with JSON lines, I would like the option to layer something deeper and closer to the metal for RPC like protobuf beneath the surface triggered by a JSON lines command or some such. I'm happy as long as those hook points exist =)

**Comment by sapphire-arches** [comment]
As discussed [on discord](https://discord.com/channels/608916579421257728/1107026299945496577/1123763391719280750) we probably want to use MIDI SYSEX to tunnel the RPC protocol (JRPC or GPRC or whatever)  to avoid driver compatibility problems.
